fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/services/MembershipIndexManager.java

```diff
@@ -0,0 +1,602 @@
+
+    private static final String INSERT_MEMBERSHIP_IN_TX =
+            "INSERT INTO membership_tx_operations" +
+            " (subject_id, property, object_id, source_id, start_time, end_time, tx_id, operation)" +
+            " VALUES (:subjectId, :property, :targetId, :sourceId, :startTime, :endTime, :txId, :operation)";
+
+    private static final String END_EXISTING_MEMBERSHIP =
+            "INSERT INTO membership_tx_operations" +
+            " (subject_id, property, object_id, source_id, start_time, end_time, tx_id, operation)" +
+            " SELECT m.subject_id, m.property, m.object_id, m.source_id, m.start_time, :endTime, :txId, :deleteOp" +
+            " FROM membership m" +
+            " WHERE m.source_id = :sourceId" +
+                " AND m.end_time = :noEndTime" +
+                " AND m.subject_id = :subjectId" +
+                " AND m.property = :property" +
+                " AND m.object_id = :objectId";
+
+    private static final String END_MEMBERSHIP_IN_TX =
+            "UPDATE membership_tx_operations" +
+            " SET end_time = :endTime" +
+            " WHERE source_id = :sourceId" +
+                " AND end_time = :noEndTime" +
+                " AND subject_id = :subjectId" +
+                " AND property = :property" +
+                " AND object_id = :objectId";
+
+    private static final String CLEAR_ENTRY_IN_TX =
+            "DELETE FROM membership_tx_operations" +
+            " WHERE source_id = :sourceId" +
+                " AND tx_id = :txId" +
+                " AND subject_id = :subjectId" +
+                " AND property = :property" +
+                " AND object_id = :objectId" +
+                " AND operation = :operation";
+
+    private static final String CLEAR_ALL_ADDED_FOR_SOURCE_IN_TX =
+            "DELETE FROM membership_tx_operations" +
+            " WHERE source_id = :sourceId" +
+                " AND tx_id = :txId" +
+                " AND operation = :addOp";
+
+    // Add "delete" entries for all existing membership from the given source, if not already deleted
+    private static final String END_EXISTING_FOR_SOURCE =
+            "INSERT INTO membership_tx_operations" +
+            " (subject_id, property, object_id, source_id, start_time, end_time, tx_id, operation)" +
+            " SELECT subject_id, property, object_id, source_id, start_time, :endTime, :txId, :deleteOp" +
+            " FROM membership m" +
+            " WHERE source_id = :sourceId" +
+                " AND end_time = :noEndTime" +
+                " AND NOT EXISTS (" +
+                    " SELECT TRUE" +
+                    " FROM membership_tx_operations mtx" +
+                    " WHERE mtx.subject_id = m.subject_id" +
+                        " AND mtx.property = m.property" +
+                        " AND mtx.object_id = m.object_id" +
+                        " AND mtx.source_id = m.source_id" +
+                        " AND mtx.operation = :deleteOp" +
```

Comment:
mainly for programmer error. When I was working on it I originally had 'delete' hardcoded in the query, but ended up with mismatches between the values in different places ('delete' vs 'deleted') which failed silently since it was valid sql. By passing in a constant it means I'll get an error if I misspell `:deleteOp` since the variable would be missing.